metrics/collectors/shared.lua

```diff
@@ -72,6 +74,30 @@ function Shared:dec(num, label_pairs)
     self.label_pairs[key] = label_pairs
 end
 
+--- Measure latency of function call
+--
+-- @param label_pairs either table with labels or function to generate labels.
+--      If function is given its called with the results of pcall.
+-- @param fn function for pcall to instrument
+-- ... - args for function fn
+function Shared:observe_latency(label_pairs, fn, ...)
+    return (function(start_time, ok, result, ...)
```

Comment:
Could you move `observe_latency_tail` into separate function?

Reply:
Why? Performance degradation?

Reply:
I think so. AFAIK luajit doesn't like closures. My recommendation initially inspired via box.atomic function. And if you could prove that box.atomic performance could be improved with inlined "tail" (as it was done here) file an issue to tarantool repository.

Reply:
Moved